interface/src/app/analyze/analysis/sampleBin.js

```diff
@@ -183,10 +197,86 @@ function($log, $cacheFactory, $q, Sample, Activity, NodeInfo, MathFuncts) {
       ).$promise;
       return pSample;
     },
+    getCachedNodeInfo: function(pk) {
+      return this.nodeCache.get(pk);
+    },
+    getNodeInfoSetPromise: function(pkArr) {
+      // Check for any pk already cached, then retrieve what's missing in
+      // bulk via the set endpoint on the node API. Return a promise and
+      // supply a callback that populates the cache when the API returns.
+      var cbSampleBin = this; // closure link to SampleBin for callbacks
+      var defer = $q.defer();
+      var cachedNodeInfoSet = [];
+
+      var uncachedPkArr = pkArr.reduce(function(acc, val) {
+        var cachedVal = cbSampleBin.getCachedNodeInfo(val);
+        if (!cachedVal) {
+          // cache does not have this pk, so keep it in our accumulator
+          acc.push(val);
+        } else {
+          cachedNodeInfoSet.push(cachedVal);
+        }
+        return acc;
+      }, []);
+      if (uncachedPkArr.length === 0) {
+        // we've got everything cached already; return before calling the API
+        defer.resolve(cachedNodeInfoSet);
+        return defer.promise;
+      }
+      NodeInfoSet.get(
+        {ids: uncachedPkArr.join(';')},
+        function success(responseObject) {
+          var i;
+          var nodeInfoArr = responseObject.objects;
+          for (i = 0; i < nodeInfoArr.length; i++) {
+            // populate the cache with what came back
+            cbSampleBin.nodeCache.put(nodeInfoArr[i].id, nodeInfoArr[i]);
+          }
+          defer.resolve(cachedNodeInfoSet.concat(nodeInfoArr));
+        },
+        function error(httpResponse) {
+          $log.error(errGen('Error retrieving NodeInfoSet', httpResponse));
+          defer.reject(httpResponse);
+        }
+      );
+
+      return defer.promise;
+    },
+    getNodeInfoPromise: function(pk) {
+      // Retrieve NodeInfo data for node id=pk from a cache, if available,
+      // returning a promise that is already fulfilled. If node `pk` is not
+      // cached, use the API to get it and add it to the cache.
+      var cbSampleBin = this; // closure link to SampleBin for callbacks
+      var defer = $q.defer();
+
+      // check the cache first and return what's there, if found
+      var cachedNode = this.getCachedNodeInfo(pk);
+      if (cachedNode) {
+        defer.resolve(cachedNode);
+        return defer.promise;
+      }
+
+      // we didn't return above, so pk is not in the cache => fetch it
+      NodeInfo.get({id: pk},
+        function success(responseObject) {
+          cbSampleBin.nodeCache.put(pk, responseObject);
+          defer.resolve(responseObject);
+        },
+        function error(httpResponse) {
+          // TODO log an error message (see Issue #79)
+          $log.error(errGen('Error retrieving NodeInfo', httpResponse));
+          defer.reject(httpResponse);
+        }
+      );
+      return defer.promise;
+    },
 
     _getIDs: function(val, i, arr) {
```

Comment:
Good call. I cleaned up one or two other places with this problem, but missed this one.